### PR TITLE
Add Basic Integration Tests for redesigned HTTP

### PR DESCRIPTION
**Pull request by sukhmanm**
*Description of changes:* Adding tests for basic HTTP requests (GET, HEAD, POST, and PUT). 

TBD: Additional tests + edge cases.

By submitting this pull request, I confirm that my contribution is made under the terms of the MIT license.

### CODE REVIEW

File: integration-test/http/http_system_test.c [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+}
+
+/* ========================== Test Cases ============================ */
+
+/**
+ * @brief Sends a GET request synchronously and verifies the result.
+ */
+void test_HTTP_GET_Request( void )
+{
+    sendHttpRequest( &transportInterface,
```
Comment (sarenameas) [inline integration-test/http/http_system_test.c:317]: For stabilization of the tests, I think we should also retry failed requests with exponential back off.

File: integration-test/http/http_system_test.c [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+ */
+#define REQUEST_BODY_LENGTH    ( sizeof( REQUEST_BODY ) - 1 )
+
+/**
+ * @brief Represents the OpenSSL context used for TLS session with the broker
+ * for tests.
+ */
+static NetworkContext_t networkContext;
+
+/* The transport layer interface used by the HTTP Client library. */
```
Comment (sarenameas) [inline integration-test/http/http_system_test.c:94]:
```suggestion
/**
 * The transport layer interface used by the HTTP Client library. 
 */
```

File: integration-test/http/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        ${stest_name} PRIVATE
+        CLIENT_CERT_PATH="${CLIENT_CERT_PATH}"
+    )
+endif()
+if(CLIENT_PRIVATE_KEY_PATH)
+    target_compile_definitions(
+        ${stest_name} PRIVATE
+        CLIENT_PRIVATE_KEY_PATH="${CLIENT_PRIVATE_KEY_PATH}"
+    )
+endif()
```
Comment (sarenameas) [inline integration-test/http/CMakeLists.txt:88]: Also newline lol

File: integration-test/http/test_config.h [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+ */
+
+/**
+ * @brief ALPN protocol name to be sent as part of the message sent to AWS IoT Core.
+ *
+ * @note Only define when using AWS IoT Core. When using ALPN, port 443 must be used.
+ *
+ * #define IOT_CORE_ALPN_PROTOCOL_NAME       "\x0ex-amzn-http-ca"
+ */
+#define IOT_CORE_ALPN_PROTOCOL_NAME       "\x0ex-amzn-http-ca"
```
Comment (sarenameas) [inline integration-test/http/test_config.h:98]: Is this used for the test?

File: integration-test/http/http_system_test.c [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+    RetryUtilsStatus_t retryUtilsStatus = RetryUtilsSuccess;
+    /* Struct containing the next backoff time. */
+    RetryUtilsParams_t reconnectParams;
+    /* Status returned by OpenSSL transport implementation. */
+    OpensslStatus_t opensslStatus;
+
+    /* Reset or initialize file-scoped global variables. */
+    ( void ) memset( &opensslCredentials, 0, sizeof( opensslCredentials ) );
+    opensslCredentials.pRootCaPath = ROOT_CA_CERT_PATH;
+    opensslCredentials.pClientCertPath = CLIENT_CERT_PATH;
```
Comment (sarenameas) [inline integration-test/http/http_system_test.c]: Let's delete lines 157-160 because we are not doing the integration tests against AWS IoT. AWS IoT only supports the POST method.

Comment (sukhmanm) [reply integration-test/http/http_system_test.c]: Removed them in the most recent commit :)

File: integration-test/http/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        ${stest_name} PRIVATE
+            ROOT_CA_CERT_PATH="${ROOT_CA_CERT_PATH}"
+    )
+endif()
+if(HTTPS_PORT)
+    target_compile_definitions(
+        ${stest_name} PRIVATE
+            HTTPS_PORT=${HTTPS_PORT}
+    )
+endif()
```
Comment (sarenameas) [inline integration-test/http/CMakeLists.txt:94]: Can you please add a new line

File: integration-test/http/http_system_test.c [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+    {
+        /* Send the request and receive the response. */
+        httpStatus = HTTPClient_Send( pTransportInterface,
+                                      &requestHeaders,
+                                      ( uint8_t * ) REQUEST_BODY,
+                                      REQUEST_BODY_LENGTH,
+                                      &response,
+                                      0 );
+
+        if( httpStatus != HTTP_SUCCESS )
```
Comment (sarenameas) [inline integration-test/http/http_system_test.c:265]: We only want to retry for HTTP_NETWORK_ERROR

File: integration-test/http/http_system_test.c [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+                                  &response,
+                                  0 );
+
+    /* If a network error is found, retry sending request for a count of MAX_RETRY_COUNT.
+     *  Proceed when MAX_RETRY_COUNT has been hit, or when a successful request is sent,
+     *  whichever comes first. */
+    if( ( httpStatus == HTTP_NETWORK_ERROR ) && ( retryCount < MAX_RETRY_COUNT ) )
+    {
+        LogDebug( ( "A network error has occured, retrying request." ) );
+        resetTest();
```
Comment (sarenameas) [inline integration-test/http/http_system_test.c:266]: If you reset the test does it jump to setUp()? If it does jump tp setUp, then won't this function be called again with retryCount set to 0?

Comment (sarenameas) [reply integration-test/http/http_system_test.c:266]: Oh I see that resetTest() just calls teardown then setup only.
```
void resetTest(void)
{
  tearDown();
  setUp();
}
```
so never mind.

File: integration-test/http/http_system_test.c [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+                    "Response Status:\n%u\n"
+                    "Response Body:\n%.*s\n",
+                    ( int32_t ) SERVER_HOST_LENGTH, SERVER_HOST,
+                    ( int32_t ) requestInfo.pathLen, requestInfo.pPath,
+                    ( int32_t ) response.headersLen, response.pHeaders,
+                    response.statusCode,
+                    ( int32_t ) response.bodyLen, response.pBody ) );
+
+        /* Close connection if a "Connection: close" header is found */
+        if( response.respFlags == HTTP_RESPONSE_CONNECTION_CLOSE_FLAG )
```
Comment (sarenameas) [inline integration-test/http/http_system_test.c:284]: I think this is redundant because tearDown will close the connection anyways after the test?

Comment (sukhmanm) [reply integration-test/http/http_system_test.c:284]: That's true -- removed it, thanks.

File: integration-test/http/http_system_test.c [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+{
+    /* Configurations of the initial request headers that are passed to
+     * #HTTPClient_InitializeRequestHeaders. */
+    HTTPRequestInfo_t requestInfo;
+    /* Represents a response returned from an HTTP server. */
+    HTTPResponse_t response;
+    /* Represents header data that will be sent in an HTTP request. */
+    HTTPRequestHeaders_t requestHeaders;
+    /* Status returned by methods in HTTP Client Library API. */
+    HTTPStatus_t httpStatus;
```
Comment (sarenameas) [inline integration-test/http/http_system_test.c:206]: We typically always initialize local variables.

**Review by sarenameas — APPROVED** [review]

File: integration-test/http/http_system_test.c [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+                                         TRANSPORT_SEND_RECV_TIMEOUT_MS );
+
+        if( opensslStatus != OPENSSL_SUCCESS )
+        {
+            retryUtilsStatus = RetryUtils_BackoffAndSleep( &reconnectParams );
+        }
+
+        if( retryUtilsStatus == RetryUtilsRetriesExhausted )
+        {
+            LogDebug( ( "Connection to the server failed, all attempts exhausted." ) );
```
Comment (aggarg) [inline integration-test/http/http_system_test.c:180]: LogError or Debug?

File: integration-test/http/http_system_test.c [hunk 12]
```diff
@@ -0,0 +1,362 @@
...
+                                      &requestHeaders,
+                                      ( uint8_t * ) REQUEST_BODY,
+                                      REQUEST_BODY_LENGTH,
+                                      &response,
+                                      0 );
+
+        if( httpStatus == HTTP_NETWORK_ERROR )
+        {
+            LogDebug( ( "A network error has occured, retrying request." ) );
+            resetTest();
```
Comment (aggarg) [inline integration-test/http/http_system_test.c:269]: Is this a Unity API? If so, has this been tested?

File: integration-test/http/http_system_test.c (hunk 11 above)
Comment (sukhmanm) [reply integration-test/http/http_system_test.c:180]: Should be LogError, thanks for pointing that out.

File: integration-test/http/http_system_test.c (hunk 12 above)
Comment (sukhmanm) [reply integration-test/http/http_system_test.c:269]: `resetTest()` gets generated with the test runner -- the function itself just calls `tearDown()` and `setUp()`, which in this case are responsible for disconnecting and reconnecting the network. I have only tested it locally with some false network errors, but it had been working as expected, and the CMock docs clarify that it can be called from within tests.

**Review by sarenameas — APPROVED** [review]

**Review by aggarg — APPROVED** [review]
